### PR TITLE
Support XCTest, Expecta and Nimble assertions in Kiwi specs

### PR DESCRIPTION
**Pull request by sharplet**
This change allows the use of XCTest assertions in specs. This should hopefully include other matcher libraries implemented in terms of XCTest assertions!

To be clear, I'm thinking of adding some more functional specs in this PR to test out whether this will enable support for Expecta or Nimble matchers.

This now works:

```objc
#import <Kiwi/Kiwi.h>

SPEC_BEGIN(KWObjCXCTestAssertionTests)

describe(@"XCTest assertions in Objective C", ^{
    it(@"supports XCTAssert", ^{
        XCTAssert(1 + 1 == 3); // <- fails with a real test failure instead of an exception or warning!
    });
});

SPEC_END
```

*****

- Within `+buildExampleGroups`, defines a shadow `self` pointer that is of the type of the test case instance.
- Unrecognized selectors sent to the test case class are forwarded to
the current example, and then to the current test case instance.
- Adds a test for the use of XCTest assertions within specs.

This should possibly be considered a breaking change, because any users depending on the _type_ of `self` being their test case class (and for example invoking class methods on `KWSpec` different in a category) will now get compiler warnings (that would be compiler errors in the case of `-Werror`).

However:

- This will just generate compiler warnings that could safely be ignored, as the tests themselves won't fail because `self` is still actually the class object at runtime!
- It should be fairly straightforward to redefine any of those class methods as instance methods.
- While valid code, it's not an encouraged or documented way of writing test helper methods.

To make this not a breaking change (at compile time), a workaround could be for this PR to change the type of `self` to `id`. This is probably even more accurate! However the convenience of the static type declaration is useful. I just don't know if this will be an issue in practice for Kiwi users.

### CODE REVIEW

File: Tests/KWObjCXCTestAssertionTests.m [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#import <Kiwi/Kiwi.h>
+
+SPEC_BEGIN(KWObjCXCTestAssertionTests)
+
+describe(@"XCTest assertions in Objective C", ^{
+    it(@"supports XCTAssert", ^{
+        XCTAssert(1 + 1 == 2);
```
Comment (sharplet) [inline Tests/KWObjCXCTestAssertionTests.m]: This test passes. In fact, it actually would have passed without any of the other changes in this PR. 0_o

The important thing is actually not that this test passes, but that when it fails it fails _in the right way_. I don't actually know how to verify that. ¯\\\_(ツ)\_/¯

**Comment by supermarin** [comment]
:+1: for the code.
Will give it a bit of thought on how to test your assertion

**Comment by sharplet** [comment]
I've just pushed a couple more commits adding explicit support for both Expecta and Nimble matchers!

**Comment by sharplet** [comment]
Been thinking on testing that assertions behave correctly, and I could potentially follow a similar approach to this: https://github.com/kiwi-bdd/Kiwi/blob/master/test_suite_configuration.rb.

Basically:

- Create a separate test bundle for testing matcher failures, and move the XCTest, Expecta and Nimble specs there.
- Write a script that matches on the output of running that test bundle.
- Integrate the script into CI.

However, it would be a lot nicer to have it properly integrated into the existing test suite.

**Comment by sharplet** [comment]
Yo @supermarin, came up with something for testing that assertions occur: a77c3e7.

**Comment by supermarin** [comment]
:shipit:

**Comment by sharplet** [comment]
I looked into trying to do write the same kind of test in the Swift specs, but we don't even support full Kiwi matchers in Swift yet. And it's working really well in Objective-C, so I don't think it's necessary to duplicate the effort in this PR.

:boat:

**Comment by aspcartman** [comment]
It's not working for me

```objc
#import <Kiwi/Kiwi.h>
#import <Expecta/Expecta.h>

SPEC_BEGIN(TestSpec)

describe(@"Just a test", ^{
    it(@"Works",^{
	    expect(1).to.equal(1);
    });
});

SPEC_END
```

It errors on `Expecta/EXPMatchers+beNil.h` on `(3,21): expected ')'`

**Pull request merged**